Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -2898,6 +2902,109 @@ bool Element::has_paint_containment() const
     return false;
 }
 
+int Element::number_of_owned_list_items() const
+{
+    auto number_of_owned_li_elements = 0;
+    for_each_child_of_type<DOM::Element>([&](auto& child) {
+        if (child.list_owner() == this) {
+            number_of_owned_li_elements++;
+        }
+        return IterationDecision::Continue;
+    });
+    return number_of_owned_li_elements;
+}
+
+// https://html.spec.whatwg.org/#list-owner
+Element const* Element::list_owner() const
+{
+    // Any element whose computed value of 'display' is 'list-item' has a list owner, which is determined as follows:
+    if (!computed_properties()->display().is_list_item())
+        return nullptr;
+
+    // 1. If the element is not being rendered, return null; the element has no list owner.
+    if (!layout_node())
+        return nullptr;
+
+    // 2. Let ancestor be the element's parent.
+    auto const* ancestor = parent_element();
+
+    // 3. If the element has an ol, ul, or menu ancestor, set ancestor to the closest such ancestor element.
+    for (auto const* node = ancestor; node; node = node->parent_element()) {
+        if (is<HTML::HTMLOListElement>(node) || is<HTML::HTMLUListElement>(node) || is<HTML::HTMLMenuElement>(node)) {
+            ancestor = node;
+            break;
+        }
+    }
+
+    // 4. Return the closest inclusive ancestor of ancestor that produces a CSS box.
+    for (; ancestor; ancestor = ancestor->parent_element()) {
+        if (ancestor->layout_node()) {
```

Comment:
Is the presence of a layout node the same thing as "producing a CSS box"? Or should we query some property of the layout node (or subclass of it)?

Reply:
Apparently, CSS box is short for CSS layout box. I updated the code accordingly so that the presence of a `paintable_box` is checked.